# Ensemble Learning

## Initial Imports

In [56]:
import warnings
warnings.filterwarnings('ignore')

In [57]:
import numpy as np
import pandas as pd
from pathlib import Path
from collections import Counter

In [58]:
from sklearn.metrics import balanced_accuracy_score
from sklearn.metrics import confusion_matrix
from imblearn.metrics import classification_report_imbalanced

## Read the CSV and Perform Basic Data Cleaning

In [59]:
# Load the data
file_path = Path('Resources/LoanStats_2019Q1.csv')
df = pd.read_csv(file_path)

# Preview the data
df.head()

loan_amnt  int_rate  installment home_ownership  annual_inc  \
0    10500.0    0.1719       375.35           RENT     66000.0   
1    25000.0    0.2000       929.09       MORTGAGE    105000.0   
2    20000.0    0.2000       529.88       MORTGAGE     56000.0   
3    10000.0    0.1640       353.55           RENT     92000.0   
4    22000.0    0.1474       520.39       MORTGAGE     52000.0   

  verification_status   issue_d loan_status pymnt_plan    dti  ...  \
0     Source Verified  Mar-2019    low_risk          n  27.24  ...   
1            Verified  Mar-2019    low_risk          n  20.23  ...   
2            Verified  Mar-2019    low_risk          n  24.26  ...   
3            Verified  Mar-2019    low_risk          n  31.44  ...   
4        Not Verified  Mar-2019    low_risk          n  18.76  ...   

   pct_tl_nvr_dlq  percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  \
0            85.7             100.0                   0.0        0.0   
1            91.2              50.0                   1.0        0.0   
2            66.7              50.0                   0.0        0.0   
3           100.0              50.0                   1.0        0.0   
4           100.0               0.0                   0.0        0.0   

   tot_hi_cred_lim  total_bal_ex_mort total_bc_limit  \
0          65687.0            38199.0         2000.0   
1         271427.0            60641.0        41200.0   
2          60644.0            45684.0         7500.0   
3          99506.0            68784.0        19700.0   
4         219750.0            25919.0        27600.0   

   total_il_high_credit_limit  hardship_flag  debt_settlement_flag  
0                     61987.0              N                     N  
1                     49197.0              N                     N  
2                     43144.0              N                     N  
3                     76506.0              N                     N  
4                     20000.0              N                     N  

[5 rows x 86 columns]

## Split the Data into Training and Testing

In [60]:
# Create our features
X = df.copy()
X.drop("loan_status", axis=1, inplace=True)
X = pd.get_dummies(X, columns=["home_ownership", "verification_status", "pymnt_plan", "hardship_flag", "debt_settlement_flag", "initial_list_status", "issue_d", "next_pymnt_d", "application_type"])
X.head()

loan_amnt  int_rate  installment  annual_inc    dti  delinq_2yrs  \
0    10500.0    0.1719       375.35     66000.0  27.24          0.0   
1    25000.0    0.2000       929.09    105000.0  20.23          0.0   
2    20000.0    0.2000       529.88     56000.0  24.26          0.0   
3    10000.0    0.1640       353.55     92000.0  31.44          0.0   
4    22000.0    0.1474       520.39     52000.0  18.76          0.0   

   inq_last_6mths  open_acc  pub_rec  revol_bal  ...  debt_settlement_flag_N  \
0             0.0       8.0      0.0     1609.0  ...                       1   
1             0.0      17.0      1.0    18368.0  ...                       1   
2             0.0       8.0      0.0    13247.0  ...                       1   
3             1.0      10.0      1.0    17996.0  ...                       1   
4             1.0      14.0      0.0     9091.0  ...                       1   

   initial_list_status_f  initial_list_status_w  issue_d_Feb-2019  \
0                      0                      1                 0   
1                      0                      1                 0   
2                      0                      1                 0   
3                      0                      1                 0   
4                      0                      1                 0   

   issue_d_Jan-2019  issue_d_Mar-2019  next_pymnt_d_Apr-2019  \
0                 0                 1                      0   
1                 0                 1                      0   
2                 0                 1                      0   
3                 0                 1                      0   
4                 0                 1                      0   

   next_pymnt_d_May-2019  application_type_Individual  \
0                      1                            1   
1                      1                            1   
2                      1                            1   
3                      1                            1   
4                      1                            1   

   application_type_Joint App  
0                           0  
1                           0  
2                           0  
3                           0  
4                           0  

[5 rows x 95 columns]

In [61]:
# Create our target
y = df.loc[:, ["loan_status"]]
y.head()

loan_status
0    low_risk
1    low_risk
2    low_risk
3    low_risk
4    low_risk

In [62]:
X.describe()

loan_amnt      int_rate   installment    annual_inc           dti  \
count  68817.000000  68817.000000  68817.000000  6.881700e+04  68817.000000   
mean   16677.594562      0.127718    480.652863  8.821371e+04     21.778153   
std    10277.348590      0.048130    288.062432  1.155800e+05     20.199244   
min     1000.000000      0.060000     30.890000  4.000000e+01      0.000000   
25%     9000.000000      0.088100    265.730000  5.000000e+04     13.890000   
50%    15000.000000      0.118000    404.560000  7.300000e+04     19.760000   
75%    24000.000000      0.155700    648.100000  1.040000e+05     26.660000   
max    40000.000000      0.308400   1676.230000  8.797500e+06    999.000000   

        delinq_2yrs  inq_last_6mths      open_acc       pub_rec  \
count  68817.000000    68817.000000  68817.000000  68817.000000   
mean       0.217766        0.497697     12.587340      0.126030   
std        0.718367        0.758122      6.022869      0.336797   
min        0.000000        0.000000      2.000000      0.000000   
25%        0.000000        0.000000      8.000000      0.000000   
50%        0.000000        0.000000     11.000000      0.000000   
75%        0.000000        1.000000     16.000000      0.000000   
max       18.000000        5.000000     72.000000      4.000000   

           revol_bal  ...  debt_settlement_flag_N  initial_list_status_f  \
count   68817.000000  ...                 68817.0           68817.000000   
mean    17604.142828  ...                     1.0               0.123879   
std     21835.880400  ...                     0.0               0.329446   
min         0.000000  ...                     1.0               0.000000   
25%      6293.000000  ...                     1.0               0.000000   
50%     12068.000000  ...                     1.0               0.000000   
75%     21735.000000  ...                     1.0               0.000000   
max    587191.000000  ...                     1.0               1.000000   

       initial_list_status_w  issue_d_Feb-2019  issue_d_Jan-2019  \
count           68817.000000      68817.000000      68817.000000   
mean                0.876121          0.371696          0.451066   
std                 0.329446          0.483261          0.497603   
min                 0.000000          0.000000          0.000000   
25%                 1.000000          0.000000          0.000000   
50%                 1.000000          0.000000          0.000000   
75%                 1.000000          1.000000          1.000000   
max                 1.000000          1.000000          1.000000   

       issue_d_Mar-2019  next_pymnt_d_Apr-2019  next_pymnt_d_May-2019  \
count      68817.000000           68817.000000           68817.000000   
mean           0.177238               0.383161               0.616839   
std            0.381873               0.486161               0.486161   
min            0.000000               0.000000               0.000000   
25%            0.000000               0.000000               0.000000   
50%            0.000000               0.000000               1.000000   
75%            0.000000               1.000000               1.000000   
max            1.000000               1.000000               1.000000   

       application_type_Individual  application_type_Joint App  
count                 68817.000000                68817.000000  
mean                      0.860340                    0.139660  
std                       0.346637                    0.346637  
min                       0.000000                    0.000000  
25%                       1.000000                    0.000000  
50%                       1.000000                    0.000000  
75%                       1.000000                    0.000000  
max                       1.000000                    1.000000  

[8 rows x 95 columns]

In [63]:
# Check the balance of our target values
y['loan_status'].value_counts()

low_risk     68470
high_risk      347
Name: loan_status, dtype: int64

In [64]:
# Split the X and y into X_train, X_test, y_train, y_test
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, 
                                                    y, 
                                                    random_state=1, 
                                                    stratify=y)

## Data Pre-Processing

Scale the training and testing data using the `StandardScaler` from `sklearn`. Remember that when scaling the data, you only scale the features data (`X_train` and `X_testing`).

In [65]:
# Create the StandardScaler instance
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()

In [66]:
# Fit the Standard Scaler with the training data
# When fitting scaling functions, only train on the training dataset
X_scaler = scaler.fit(X_train)

In [67]:
# Scale the training and testing data
X_train = X_scaler.transform(X_train)
X_test = X_scaler.transform(X_test)

## Ensemble Learners

In this section, you will compare two ensemble algorithms to determine which algorithm results in the best performance. You will train a Balanced Random Forest Classifier and an Easy Ensemble classifier . For each algorithm, be sure to complete the folliowing steps:

1. Train the model using the training data. 
2. Calculate the balanced accuracy score from sklearn.metrics.
3. Display the confusion matrix from sklearn.metrics.
4. Generate a classication report using the `imbalanced_classification_report` from imbalanced-learn.
5. For the Balanced Random Forest Classifier only, print the feature importance sorted in descending order (most important feature to least important) along with the feature score

Note: Use a random state of 1 for each algorithm to ensure consistency between tests

### Balanced Random Forest Classifier

In [68]:
# Resample the training data with the BalancedRandomForestClassifier
from imblearn.ensemble import BalancedRandomForestClassifier
brf = BalancedRandomForestClassifier(random_state=1)
brf.fit(X_train, y_train)

BalancedRandomForestClassifier(random_state=1)

In [69]:
# Calculated the balanced accuracy score
from sklearn.metrics import balanced_accuracy_score
y_pred = brf.predict(X_test)
balanced_accuracy_score(y_test, y_pred)

0.778099076994976

In [70]:
# Display the confusion matrix
from sklearn.metrics import confusion_matrix
confusion_matrix(y_test, y_pred)

array([[   58,    29],
       [ 1891, 15227]])

In [71]:
# Print the imbalanced classification report
from imblearn.metrics import classification_report_imbalanced
print(classification_report_imbalanced(y_test, y_pred))

                   pre       rec       spe        f1       geo       iba       sup

  high_risk       0.03      0.67      0.89      0.06      0.77      0.58        87
   low_risk       1.00      0.89      0.67      0.94      0.77      0.61     17118

avg / total       0.99      0.89      0.67      0.94      0.77      0.61     17205



In [72]:
# List the features sorted in descending order by feature importance
importances = brf.feature_importances_
sorted(zip(brf.feature_importances_, X.columns), reverse=True)

[(0.06813071243290279, 'total_rec_prncp'),
 (0.06580308434742649, 'last_pymnt_amnt'),
 (0.06396941048680489, 'total_pymnt_inv'),
 (0.05191870234468143, 'total_rec_int'),
 (0.04541378226698817, 'total_pymnt'),
 (0.023975907183493763, 'int_rate'),
 (0.019361782313250753, 'max_bal_bc'),
 (0.018619759728779602, 'mths_since_recent_inq'),
 (0.018202719129787635, 'installment'),
 (0.018006720787573167, 'annual_inc'),
 (0.016776951622973747, 'out_prncp'),
 (0.016689420485935934, 'revol_bal'),
 (0.016315530644942036, 'total_bal_ex_mort'),
 (0.016271693615257207, 'il_util'),
 (0.01611515198161295, 'dti'),
 (0.016081357457945804, 'bc_util'),
 (0.015887840128831023, 'issue_d_Jan-2019'),
 (0.015752493742845906, 'avg_cur_bal'),
 (0.015396128243538578, 'total_bc_limit'),
 (0.014866251162986259, 'mo_sin_old_il_acct'),
 (0.014823283152547655, 'out_prncp_inv'),
 (0.014164127683518204, 'mo_sin_old_rev_tl_op'),
 (0.014142924623164127, 'all_util'),
 (0.013938373648565296, 'tot_hi_cred_lim'),
 (0.0138756686

### Easy Ensemble Classifier

In [73]:
# Train the Classifier
from imblearn.ensemble import EasyEnsembleClassifier
eec = EasyEnsembleClassifier(random_state=1)
eec.fit(X_train, y_train)

EasyEnsembleClassifier(random_state=1)

In [74]:
# Calculated the balanced accuracy score
from sklearn.metrics import balanced_accuracy_score
y_pred = eec.predict(X_test)
balanced_accuracy_score(y_test, y_pred)

0.9259823295502616

In [75]:
# Display the confusion matrix
from sklearn.metrics import confusion_matrix
confusion_matrix(y_test, y_pred)

array([[   79,     8],
       [  960, 16158]])

In [76]:
# Print the imbalanced classification report
from imblearn.metrics import classification_report_imbalanced
print(classification_report_imbalanced(y_test, y_pred))

                   pre       rec       spe        f1       geo       iba       sup

  high_risk       0.08      0.91      0.94      0.14      0.93      0.85        87
   low_risk       1.00      0.94      0.91      0.97      0.93      0.86     17118

avg / total       0.99      0.94      0.91      0.97      0.93      0.86     17205



### Final Questions

1. Which model had the best balanced accuracy score?

    EasyEnsembleClassifier

2. Which model had the best recall score?

    EasyEnsembleClassifier

3. Which model had the best geometric mean score?

    EasyEnsembleClassifier

4. What are the top three features?

    (0.06813071243290279, 'total_rec_prncp')
    (0.06580308434742649, 'last_pymnt_amnt')
    (0.06396941048680489, 'total_pymnt_inv')
